sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
     std::unique_ptr<sstable_set_impl> _impl;
-    schema_ptr _schema;
+    std::unordered_map<utils::UUID, std::unordered_set<shared_sstable>> _all_runs;
+    std::unordered_map<shared_sstable, unsigned> _times_added;
+    std::set<shared_sstable> _all;
+public:
+    versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl);
+    void insert(shared_sstable sst);
+    void erase(shared_sstable sst);
+    std::vector<shared_sstable> select(const dht::partition_range& range) const;
+    std::unordered_set<shared_sstable> select_by_run_id(utils::UUID run_id) const;
+    const std::set<shared_sstable>& all() const;
+    void remove(shared_sstable sst);
+    std::unique_ptr<incremental_selector_impl> make_incremental_selector();
+};
+
+class sstable_set_version;
+
+class sstable_set_version_reference {
+    mutable sstable_set_version* _p = nullptr;
+public:
+    ~sstable_set_version_reference();
+    sstable_set_version_reference() = default;
+    sstable_set_version_reference(sstable_set_version* p);
+    sstable_set_version_reference(const sstable_set_version_reference& ref);
+    sstable_set_version_reference(sstable_set_version_reference&&);
+    sstable_set_version_reference& operator=(const sstable_set_version_reference& x);
+    sstable_set_version_reference& operator=(sstable_set_version_reference&&);
+    sstable_set_version& operator*() const noexcept { return *_p; }
+    sstable_set_version* operator->() const noexcept { return _p; }
+    sstable_set_version* get() const noexcept { return _p; }
+    explicit operator bool() const noexcept { return bool(_p); }
+};
+
+class sstable_list {
+    sstable_set_version_reference _version;
+public:
+    sstable_list(std::unique_ptr<sstable_set_impl> impl, schema_ptr s);
+    sstable_list(const sstable_list& sstl);
+    sstable_list(sstable_list&& sstl) noexcept;
+    sstable_list& operator=(const sstable_list& sstl);
+    sstable_list& operator=(sstable_list&& sstl) noexcept;
+public:
+    class const_iterator : public std::iterator<std::forward_iterator_tag, const shared_sstable> {
```

Comment:
Minor: std::iterator was deprecated in C++17.

Reply:
Ok, it's simple change, so I'll fix it even if it's minor